src/YoutubeAtomPlayer.tsx

```diff
@@ -340,6 +353,35 @@ export const YoutubeAtomPlayer = ({
         ],
     );
 
+    /**
+     * Pause the current video when another video is played on the same page
+     */
+    const handlePauseVideo = (
+        event: CustomEventInit<PlayCustomEventDetail>,
+    ): void => {
+        if (event instanceof CustomEvent) {
+            if (event.detail.videoId !== videoId) {
+                const playerStatePromise = player.current?.getPlayerState();
+                playerStatePromise?.then((state) => {
+                    if (state === YT.PlayerState.PLAYING) {
+                        player.current?.pauseVideo();
+                    }
+                });
+            }
+        }
+    };
+
+    /**
+     * add listener for custom play event
+     * remove listener in clean up function
+     */
+    useEffect(() => {
+        document.addEventListener(customPlayEventName, handlePauseVideo);
+        return () => {
+            document.removeEventListener(customPlayEventName, handlePauseVideo);
+        };
+    }, []);
```

Comment:
Normally you'd need to add any variable used inside the `useEffect` to its dependency array i.e. `handlePauseVideo` should be in the dependency array.

However I think we can avoid this by moving the function creation of `handlePauseVideo` inside this `useEffect`.

Because this `useEffect` will only run once then we will only create it once and the listeners too.

Reply:
Thank @arelra for your patience and explanation :) If we move the function creation into the useeffect, do we not also need to add function dependancies to the dependancy array?

Reply:
Great catch! Yes you're right.

I think we would need the dependency `videoId`, but not `player.current` however as you are [not supposed to pass refs](https://medium.com/@teh_builder/ref-objects-inside-useeffect-hooks-eb7c15198780) to a as a usEffect dependency

In which case we can use `useCallback` to wrap the function creation but _then_ we (in theory) lose the specific unmount `useEffect` by adding that as a dependency.

I think we should probably switch back to the two-step listener setup and unregister that the current player listeners use? That way we can guarantee conforming to all the crazy useEffect gotchas 🙃

Reply:
IIUC and the only value in the dependancy array is `videoId`, that's not going to change, so the cleanup won't be fired until the component un-mounts?  

The reason we had issues previously was that somewhere in the deps array of the larger useEffect within `YoutubeAtomPlayer` dependancies were updating, which was causing the cleanup to run sooner.

I think it's a lot easier for anyone coming to this project afresh if this all lives in one useEffect, and we're not pushing listeners to refs etc unless we really have to?

Reply:
@arelra and @joecowton1 thank you both for your great ideas. 

Having paired with Ravi, we moved the listener function back to the original useEffect. 

As Joe mentioned, most probably the dependancy on `videoId` wouldn't have caused issues in useEffect as it's a static value. But I think 1- to be on the safe side (avoiding all the crazy useEffect gotchas ) and 2- to be consistent with how the rest of the listeners have been registered and de-registered, it may be good to follow the same pattern of registering in one useEffect and de-registering in a separate useEffect?

Please let me know if you feel concerened about it? Or you think there could still be more improvements 🙂

Reply:
> the only value in the dependancy array is videoId, that's not going to change, so the cleanup won't be fired until the component un-mounts

Yep you're right! However this only works as a consequence of how that particular value is changed or not changed over the lifecycle of the component.

This may not work for other variables if they are introduced later to the component (and useEffect) so I'm a bit wary of introducing a pattern that might not work correctly in call cases for all dependencies.

The current pattern of refs for the player and player listeners is already established and I personally think thats actually easier and more deterministic than trying to grok when a useEffect might fire depending on its dependencies? 

But I'm very aware that might just be me as I have been staring at this for a while :)

Reply:
With the caveat that we've all being looking at this for a long time now, and we're probably hitting the limits of the cost-benefit of a prolonged back and forth, my current thoughts are..

> However this only works as a consequence of how that particular value is changed or not changed over the lifecycle of the component.

Is this not true of every useEffect ever? :)

> The current pattern of refs for the player and player listeners is already established and I personally think thats 
> actually easier and more deterministic than trying to grok when a useEffect might fire depending on its 
> dependencies?

I think that an individual useEffect provides a separation of concerns and is easy to understand - to me it feels as though this is separate from the player listeners for which pushing to refs is a workaround, this doesn't require that.

To me something like below is a lot less intuitive than a useEffect that adds and removes and event listener, something that's very common place in React projects everywhere. I think without context I would wonder why we needed to do this:

```javascript
      customListeners.current[customPlayEventName] = handlePauseVideo;
                
     Object.entries(customListeners.current).forEach(
                ([eventName, eventHandler]) => {
                    document.removeEventListener(eventName, eventHandler);
                },
            );
```

Anyway happy to let it to slide, that's just my take on it. It's good to throughly interrogate this things 🙂

Reply:
Valid points you're making Joe. It's also about personal preference which could look more or less intuitive. 
But one point I understood from Ravi's comment is that if we have a useEffect that is having a non-empty dependancy array even if it's a static dependancy`[videoId]`, this way, we're introducing this pattern that it's OK for this useEffect to have dependancy and make it more prone to bug in the future

Reply:
> Is this not true of every useEffect ever? :)

I get your point and its totally valid but I think in the particular case of unregistering listeners we want some strong guarantees that we will only do that on onmount. If we can avoid setting potential traps for ourselves that require deep understanding of re-renders, function recreation and useEffect gotchas I feel like that's a better place to be.

> this is separate from the player listeners

Yep I can see that. It gets subjective here! To me they are all listeners and if we register and deregister them in the same way I feel that groups them together and aids understanding.

> I think without context I would wonder why we needed to do this

Thats valid perhaps we could add more detail to the comments that are already there.

I think we can all agree though that useEffect is a pain! It has a problematic API and it doesn't help devs to [fall into the pit success](https://blog.codinghorror.com/falling-into-the-pit-of-success/).

Reply:
@joecowton1 I updated the comment about adding the listeners to refs. Are you happy with the current state of the PR now?

Reply:
Thanks @marjisound - yeah nice work 🙂